client/peerset/src/protocol_controller.rs

```diff
@@ -0,0 +1,576 @@
+	pub async fn next_action(&mut self) -> bool {
+		let action = loop {
+			let mut next_alloc_slots = Delay::new_at(self.next_periodic_alloc_slots).fuse();
+			futures::select! {
+				action = self.from_handle.next() => match action {
+					Some(action) => break action,
+					None => return false,
+				},
+				_ = next_alloc_slots => {
+					self.alloc_slots();
+					self.next_periodic_alloc_slots = Instant::now() + Duration::new(1, 0);
+				},
+			}
+		};
+		match action {
+			Action::AddReservedPeer(peer_id) => self.on_add_reserved_peer(peer_id),
+			Action::RemoveReservedPeer(peer_id) => self.on_remove_reserved_peer(peer_id),
+			Action::SetReservedPeers(peer_ids) => self.on_set_reserved_peers(peer_ids),
+			Action::SetReservedOnly(reserved_only) => self.on_set_reserved_only(reserved_only),
+			Action::AddPeer(peer_id) => self.on_add_peer(peer_id),
+			Action::RemovePeer(peer_id) => self.on_remove_peer(peer_id),
+			Action::IncomingConnection(peer_id, index) =>
+				self.on_incoming_connection(peer_id, index),
+		}
+		true
+	}
+
+	fn on_add_reserved_peer(&mut self, peer_id: PeerId) {
+		if !self.reserved_nodes.insert(peer_id) {
+			return
+		}
+		// Discount occupied slots or connect to the node.
+		match self.nodes.entry(peer_id).or_default() {
+			ConnectionState::In => self.num_in -= 1,
+			ConnectionState::Out => self.num_out -= 1,
+			ConnectionState::NotConnected => self.alloc_slots(),
+		}
+	}
+
+	fn on_remove_reserved_peer(&mut self, peer_id: PeerId) {
+		if !self.reserved_nodes.remove(&peer_id) {
+			return
+		}
+		if let Entry::Occupied(mut node) = self.nodes.entry(peer_id) {
+			if self.reserved_only && node.get().is_connected() {
+				// If we are in reserved-only mode, the removed reserved node
+				// should be disconnected.
+				// Note that we don't need to update slots, so we drop the node manually here
+				// and not via [`Peer`] interface
+				*node.get_mut() = ConnectionState::NotConnected;
+				let _ = self
+					.to_notifications
+					.unbounded_send(Message::Drop { set_id: self.set_id, peer_id: *node.key() });
+				trace!(
+					target: "peerset",
+					"Disconnecting previously reserved node {} on {:?}.",
+					node.key(), self.set_id
+				);
+			} else {
+				// Otherwise, just count occupied slots for the node.
+				match node.get() {
+					ConnectionState::In => self.num_in += 1,
+					ConnectionState::Out => self.num_out += 1,
+					ConnectionState::NotConnected => {},
+				}
+			}
+		} else {
+			error!(
+				target: "peerset",
+				"Invalid state: reserved node {} not in the list of known nodes.",
+				peer_id,
+			);
+		}
+	}
+
+	fn on_set_reserved_peers(&mut self, peer_ids: HashSet<PeerId>) {
+		// Determine the difference between the current group and the new list.
+		let to_insert = peer_ids.difference(&self.reserved_nodes).cloned().collect::<Vec<_>>();
+		let to_remove = self.reserved_nodes.difference(&peer_ids).cloned().collect::<Vec<_>>();
```

Comment:
Does the original implementation remove peers not in the new set?

Reply:
Yes